src/Agent.Worker/DiagnosticLogManager.cs

```diff
@@ -0,0 +1,247 @@
+            var configurationStore = HostContext.GetService<IConfigurationStore>();
+            AgentSettings settings = configurationStore.GetSettings();
+            int agentId = settings.AgentId;
+            string agentName = settings.AgentName;
+
+            executionContext.Debug("Creating diagnostic log environment file.");
+            string environmentFile = Path.Combine(supportFilesFolder, "environment.txt");
+            string content = GetEnvironmentContent(agentId, agentName, message.Tasks);
+            using (StreamWriter writer = File.CreateText(environmentFile)) 
+            {
+                writer.Write(content);
+            }
+
+            // Copy worker diag log files
+            List<string> workerDiagLogFiles = GetWorkerDiagLogFiles(HostContext.GetDirectory(WellKnownDirectory.Diag), jobStartTimeUtc);
+            executionContext.Debug($"Copying {workerDiagLogFiles.Count()} worker diag logs.");
+
+            foreach(string workerLogFile in workerDiagLogFiles)
+            {
+                ArgUtil.Directory(workerLogFile, nameof(workerLogFile));
+
+                string destination = Path.Combine(supportFilesFolder, Path.GetFileName(workerLogFile));
+                File.Copy(workerLogFile, destination);
+            }
+            
+            executionContext.Debug("Zipping diagnostic files.");
+
+            string buildNumber;
+            if (!message.Environment.Variables.TryGetValue(Constants.Variables.Build.Number, out buildNumber))
+            {
+                buildNumber = "UnknownBuildNumber";
+            }
+            string buildName = $"Build {buildNumber}";
+            
+            string phaseName;
+            if (!message.Environment.Variables.TryGetValue(Constants.Variables.System.PhaseDisplayName, out phaseName))
+            {
+                phaseName = "UnknownPhaseName";
+            }
+
+            // zip the files
+            string diagnosticsZipFileName = $"{buildName}-{phaseName}.zip";
+            string diagnosticsZipFilePath = Path.Combine(supportRootFolder, diagnosticsZipFileName);
+            ZipFile.CreateFromDirectory(supportFilesFolder, diagnosticsZipFilePath);
+
+            // upload the json metadata file
+            executionContext.Debug("Uploading diagnostic metadata file.");
+            string metadataFileName = $"diagnostics-{buildName}-{phaseName}.json";
+            // create the file
+            string metadataFilePath = Path.Combine(supportFilesFolder, metadataFileName);
+            using (StreamWriter writer = File.CreateText(metadataFilePath)) 
+            {
+                writer.Write(JsonUtility.ToString(new DiagnosticLogMetadata(agentName, agentId.ToString(), phaseName, diagnosticsZipFileName)));
+            }
+
+            IJobServerQueue jobServerQueue = HostContext.GetService<IJobServerQueue>();
+
+            // upload it
+            jobServerQueue.QueueFileUpload(
```

Comment:
why not just use ExecutionContext.QueueAttachFile()?

Reply:
Hadn't seen it, fixed :)